### DISCUSSION TITLE
How to get the filesystem of `$PWD`?

### DISCUSSION THREAD

**Discussion by roke-julian-lockhart** [PowerShell/PowerShell#21321] [Q&A] [answered]
In order to not write inode names with disallowed characters to filesystems which don't support most characters (like NTFS, FAT*, and/or exFAT) I would like to ensure that the `$PWD` is BTRFS. How can I do this? https://github.com/MicrosoftDocs/PowerShell-Docs/blob/1ce9e4023ee4a2ebb0af02b3108e568d34b3d09e/reference/7.4/Microsoft.PowerShell.Management/Get-PSDrive.md#get-psdrive doesn't appear to cater for this use case.

If an example of why I want this is useful, https://gitlab.com/rokejulianlockhart/rokejulianlockhart/-/blob/9a2782c3972c1436680aedd546e792bf21002baf/%09%60%7Btitle:%20%22Entities%22,%09id:%20%22s1gvg2%22%7D%60%7B.JSON5%7D.dir/%09%60%7Bname:%20%7Bshort:%20%22Beedell.RokeJulianLockhart%22,%09long:%20%7Bthird:%20%22Beedell%22,%09first:%20%22Roke%22%7D%7D,%09id:%20%22s1gvgt%22%7D%60%7B.JSON5%7D.dir/%09%60%7Btitle:%20%22Technical%22,%09id:%20%22s20cxw%22%7D%60%7B.JSON5%7D.dir/%09%60%7Btitle:%20%22Projects%22,%09id:%20%22%22%7D%60%7B.JSON5%7D.dir/%09%60%7Btitle:%20%22Configurator%22,%09id:%20%22s3wrcl%22%7D%60%7B.JSON5%7D.dir/%09%60%7Btitle:%20%22Code%22,%09id:%20%22%22%7D%60%7B.JSON5%7D.dir/%09%60%60%7Btitle:%20%22%60Set-Configuration%60%7B.PS1%7D%22,%09id:%20%22%22%7D%60%60%7B.JSON5%7D.txt.PS1#L133-135 is my current primary use case.

**Comment by rhubarb-geek-nz** [comment]
Assuming Linux due to BTRFS

GetDrives will give you the drive format

```
[System.IO.DriveInfo]::GetDrives() | Format-Table
```

you will need to get the drive by matching the root of your file system with Name

```
PS> ([System.IO.DriveInfo]::GetDrives() | Where-Object -Property Name -EQ -Value /).DriveFormat
ext3
```

Unfortunately GetPathRoot does not appear useful, it always returns / on Linux as far as I can see

```
PS> [System.IO.Path]::GetPathRoot($PWD)
/
PS> [System.IO.Path]::GetPathRoot("/tmp")
/
PS> [System.IO.Path]::GetPathRoot("/boot/firmware")
/
PS> [System.IO.Path]::GetPathRoot("/boot")
/
```

So back to POSIX and "df" and up to you to do the parsing to get the "Mounted on" value

```
PS> df $PWD
Filesystem     1K-blocks     Used Available Use% Mounted on
/dev/mmcblk0p2  60804704 18559076  39138940  33% /
PS> df  "/boot/firmware/overlays"
Filesystem     1K-blocks  Used Available Use% Mounted on
/dev/mmcblk0p1    522232 75476    446756  15% /boot/firmware
PS> df  "/run/lock/apache2"
Filesystem     1K-blocks  Used Available Use% Mounted on
tmpfs               5120    12      5108   1% /run/lock
```

On Linux GetInvalidFileNameChars just gives NUL and forward slash

```
PS> [System.IO.Path]::GetInvalidFileNameChars() | ForEach-Object { [int]$_ }
0
47
```

**Reply by roke-julian-lockhart** [reply]
> https://github.com/orgs/PowerShell/discussions/21321#discussioncomment-8726472

@rhubarb-geek-nz, do you know whether `[System.IO.Path]::GetInvalidFileNameChars()` just pulls those values from the kernel (because the kernel disallows *those* in paths) or whether it can factor in a filesystem? Sorry if I'm asking you to spoon-feed me.

**Reply by rhubarb-geek-nz** [reply]
In a POSIX system only the NUL terminator and forward slash are not permitted in a filename because the NUL is used for end-of-string in the C api and forward slash is used to indicated the directories within the full path.

However it can all turn to custard very quickly because that is assuming ASCII, once you get to MBCS and UNICODE etc you need to ensure the actual file-system can handle the characters.

The "?" and "*" are not special characters according to the kernel, only to shells and other applications, their expansion and matching are done outside of the kernel.

I would recommend being conversative with the characters you use to ensure files can be backed-up, and copied to different file systems or operating systems.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/orgs/PowerShell/discussions/21321#discussioncomment-8729949

@rhubarb-geek-nz, to my knowledge, some filesystems disallow characters too — although https://en.wikipedia.org/w/index.php?title=Filename&oldid=1193103572#Reserved_characters_and_words indeed demonstrates that it's mostly Windows's Win32 APIs which cause the biggest headaches, https://en.wikipedia.org/w/index.php?title=Filename&oldid=1193103572#Comparison_of_filename_limitations should demonstrate that the filesystems themselves can restrict them too.

**Reply by rhubarb-geek-nz** [reply]
Absolutely, the file system itself may not be able to deal with anything more than ASCII.

It isn't just Win32, it is also the various CDROM formats etc, and then what TAR and ZIP can deal with.

Hence be conservative and keep file your file name conventions simple and consistent.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/orgs/PowerShell/discussions/21321?sort=top#discussioncomment-8730069

@rhubarb-geek-nz, that's sound advice. Thank you.

Consequently, it might sound insane, but I've actually chosen https://github.com/json5/json5/blob/de344f0619bda1465a6e25c76f1c0c3dda8108d9/README.md#json5--json-for-humans as my filename format. It's made attaching attributes to files incredibly easy. I would have otherwise needed to fight with lack of support for XDG Extended Attributes — just look at the mess that is https://wiki.archlinux.org/index.php?title=Extended_attributes&oldid=798672#File_systems. The format's not transferrable to Windows and AOSP, but using Linux+BTRFS/EXT*, it's wonderful.

**Comment by roke-julian-lockhart** [comment, marked as answer]
> https://github.com/orgs/PowerShell/discussions/21321?sort=top#discussion-6342435

Per @rhubarb-geek-nz,'s damn comprehensive response at https://github.com/orgs/PowerShell/discussions/21321?sort=top#discussioncomment-8726472, I've created the answer:

```.PS1
([System.IO.DriveInfo]::GetDrives() | Where-Object -Property 'Name' -Eq -Value (df "$PWD" | awk '{if ($1 != "Filesystem") print $6}')).DriveFormat
```
